File: backend/src/services/tools/wiki.js
import db from '../../db.js';
import { removeDiacritics, tokenize } from './helpers.js';

export function search_wiki({ query }, wikiDir, userId) {
  if (!userId) return 'Không thể xác định người dùng.';
  const q = `%${query}%`;
  const rows = db.prepare(
    'SELECT title, summary, content, topics FROM wiki_entries WHERE user_id = ? AND (title LIKE ? OR summary LIKE ? OR content LIKE ?) LIMIT 5'
  ).all(userId, q, q, q);

  if (rows.length === 0) return 'Không tìm thấy kết quả nào trong Wiki.';

  return rows.map(row => {
    const topics = JSON.parse(row.topics || '[]');
    return `📄 **${row.title}** (${topics.join(', ')})\n${row.summary}\n\n${row.content.slice(0, 500)}...`;
  }).join('\n\n---\n\n');
}

export function read_page({ title }, wikiDir, userId) {
  if (!userId) return 'Không thể xác định người dùng.';
  const entry = db.prepare('SELECT * FROM wiki_entries WHERE user_id = ? AND title = ?').get(userId, title);
  if (!entry) return `Không tìm thấy trang Wiki có tiêu đề: ${title}`;
  return `📄 **${entry.title}**\n\n${entry.content}`;
}

export function listWikiTopics(wikiDir, userId) {
  if (!userId) return 'Không thể xác định người dùng.';
  const rows = db.prepare('SELECT title, topics FROM wiki_entries WHERE user_id = ? ORDER BY updated_at DESC LIMIT 20').all(userId);
  if (rows.length === 0) return 'Wiki chưa có dữ liệu.';

  const titles = rows.map(r => {
    const topics = JSON.parse(r.topics || '[]');
    return `  📄 ${r.title} [${topics.join(', ')}]`;
  });

  return ['**📚 Các trang Wiki gần đây:**', ...titles].join('\n');
}

File: backend/agent/app/api/routers/wiki.py
from __future__ import annotations

import json

from fastapi import APIRouter, HTTPException, Request

from api.services.db import get_db
from api.services.wiki_memory import resolve_user_id, save_wiki_entry

router = APIRouter(tags=["wiki"])


@router.get("/wiki")
def list_wiki(token: str | None = None) -> dict:
    user_id = resolve_user_id(f"Bearer {token}" if token else None)
    with get_db() as conn:
        conn.row_factory = None
        rows = conn.execute(
            "SELECT id, title, summary, content, topics, created_at, updated_at FROM wiki_entries WHERE user_id = ? ORDER BY updated_at DESC",
            (user_id,),
        ).fetchall()
        entries = []
        topics: dict = {}
        for row in rows:
            topics_list = json.loads(row[4]) if row[4] else ["general"]
            entry = {
                "id": row[0],
                "title": row[1],
                "summary": row[2],
                "content": row[3],
                "topics": topics_list,
                "created_at": row[5] if len(row) > 5 else None,
                "updated_at": row[6] if len(row) > 6 else None,
            }
            entries.append(entry)
            for t in topics_list:
                topics.setdefault(t, []).append(entry)
        return {"entries": entries, "topics": topics}


@router.post("/wiki")
def create_wiki_entry(request: Request, payload: dict) -> dict:
    user_id = resolve_user_id(request.headers.get("authorization"))
    entry = {
        "title": payload.get("title", ""),
        "summary": payload.get("summary", ""),
        "content": payload.get("content", ""),
        "topics": payload.get("topics", ["general"]),
    }
    result = save_wiki_entry(user_id, entry, source="manual")
    if not result:
        raise HTTPException(status_code=400, detail="Không thể lưu bài viết")
    return {"id": result["id"], "title": result["title"]}